woocommerce/woocommerce-ios
# Order Details > Shipping Label: hide order item total & subtitle when the price is unknown

Part of #2167

## Why

In a shipping label's API response, we only know the product/variation IDs or names depending on the WCS plugin version. Before the products/variations are fetched given the IDs, we do not know the price/total of the product/variation for an order item in a shipping label. For example, right now we don't fetch the missing variations for an Order in order details (it's a subtask in #2167), so that the order items with missing variations don't have a valid price/total. We currently set the missing price/total values for such order items to zero as in the before/after screenshots below, and this PR updates `AggregateOrderItem` to allow its price and total values to be `nil` so that we can hide the total & subtitle labels in this scenario.

## Changes

I took the chance to refactor `ProductDetailsCellViewModel` a bit so that it's easier to customize the logic when the price/total value is `nil`.

- Updated `AggregateOrderItem`'s `price` and `total` to be nullable
- In `AggregatedShippingLabelOrderItems`, updated to set `AggregateOrderItem`'s `price` and `total` to be `nil` if only the name is known
- In `ProductDetailsCellViewModel`, refactored to remove the private properties that were only used for public calculated properties and consolidate the logic that sets the public properties in a private initializer
  - The only behavior change is so that we set the `total` and `subtitle` strings to be empty if an item's `price`/`total` is `nil`
  - Added unit tests for the existing `ProductDetailsCellViewModel` usage for `OrderItem`/`AggregateOrderItem`/`OrderItemRefund`
- Added an Auto Layout constraint to `ProductDetailsTableViewCell`'s image view: `superview.bottom >= imageView.bottom + 8px` so that the image has some margin at the bottom when the subtitle isn't shown

## Testing

Prerequisites: the site has [WooCommerce Shipping & Tax plugin](https://docs.woocommerce.com/document/woocommerce-shipping-and-tax/).

You can set up your store to use a test credit card with instructions in p91TBi-3xD-p2.

- In core, create an order with a variation. Then create a shipping label for this order
- Launch the app, log out and in to the store without going to the Products tab
- Go to the orders tab
- Tap on the order with a variation & shipping label --> after syncing, the variation order item should not show any total or subtitle like in the example screenshot below

## Example screenshots

before | after
-- | --
<img src="https://user-images.githubusercontent.com/1945542/100717302-d431c380-33f4-11eb-8a78-f1838c2d02c6.png" width="300" /> | <img src="https://user-images.githubusercontent.com/1945542/100717784-7e115000-33f5-11eb-854e-b8b50730afd1.png" width="300" />


Update release notes:
